packages/wire/src/queue.ts

```diff
@@ -27,19 +28,83 @@ export class AsyncQueue {
   } | null = null;
   constructor() {
     this.socket = new net.Socket({});
-    this.socket.on('data', (buffer: Buffer) => {
-      debug('received %o bytes', buffer.length);
-      this.queue.push(buffer);
-      this.processQueue();
-    });
   }
-  public connect(passedOptions: { port: number; host: string }): Promise<void> {
+  public connect(passedOptions: {
+    port: number;
+    host: string;
+    ssl?: tls.ConnectionOptions | boolean;
+  }): Promise<void> {
+    const { ssl, ...connectOptions } = passedOptions;
+    const sslEnabled = ssl === true || ssl != null;
+
+    const attachDataListener = () => {
+      this.socket.on('data', (buffer: Buffer) => {
+        debug('received %o bytes', buffer.length);
+        this.queue.push(buffer);
+        this.processQueue();
+      });
+    };
+
     return new Promise((resolve) => {
       this.socket.on('connect', () => {
         debug('socket connected');
-        resolve();
+
+        if (sslEnabled) {
+          const requestSSLMessage = Buffer.allocUnsafe(8);
+          requestSSLMessage.writeInt32BE(8, 0);
+          requestSSLMessage.writeInt32BE(80877103, 4);
+
+          this.socket.write(requestSSLMessage);
+        } else {
+          attachDataListener();
+          resolve();
+        }
       });
-      this.socket.connect(passedOptions);
+
+      if (sslEnabled) {
+        this.socket.once('data', (buffer) => {
+          const responseCode = buffer.toString('utf8');
+          switch (responseCode) {
+            case 'S':
+              break;
+            case 'N':
+              this.socket.end();
+              throw new Error('The server does not support SSL connections');
+            default:
+              this.socket.end();
+              throw new Error(
+                'There was an error establishing an SSL connection',
+              );
```

Comment:
Can this message be moved here?
https://github.com/adelsz/pgtyped/blob/master/packages/wire/src/messages.ts

Reply:
So one issue I saw with doing that is I'm not sure how to handle multiple indicator responses with how you have everything setup.

Reply:
Ah, ok. Looking at the docs [here](https://www.postgresql.org/docs/9.3/protocol-flow.html#AEN99727), it does seem like the N/S response isn't a standard pgsql protocol message, so I guess it is ok to leave it as is.